src/Core/Domain/SpecificPrice/Command/AddSpecificPriceCommand.php

```diff
@@ -0,0 +1,382 @@
+
+    /**
+     * @var Number
+     */
+    private $price;
+
+    /**
+     * @var int
+     */
+    private $fromQuantity;
+
+    /**
+     * @var int|null
+     */
+    private $shopGroupId;
+
+    /**
+     * @var int|null
+     */
+    private $shopId;
+
+    /**
+     * @var int|null
+     */
+    private $cartId;
+
+    /**
+     * @var int|null
+     */
+    private $productAttributeId;
+
+    /**
+     * @var int|null
+     */
+    private $currencyId;
+
+    /**
+     * @var int|null
+     */
+    private $cartRuleId;
+
+    /**
+     * @var int|null
+     */
+    private $countryId;
+
+    /**
+     * @var int|null
+     */
+    private $groupId;
+
+    /**
+     * @var int|null
+     */
+    private $customerId;
+
+    /**
+     * @var DateTime|null
+     */
+    private $dateTimeFrom;
+
+    /**
+     * @var DateTime|null
+     */
+    private $dateTimeTo;
+
+    /**
+     * @param int $productId
+     * @param float $price
+     * @param int $fromQuantity
+     * @param string $reductionType
+     * @param float $reductionValue
+     * @param bool $includeTax
+     *
+     * @throws DomainConstraintException
+     */
+    public function __construct(
+        int $productId,
+        string $reductionType,
+        float $reductionValue,
+        bool $includeTax,
+        float $price,
+        int $fromQuantity
+    ) {
+        $this->productId = new ProductId($productId);
+        $this->reduction = new Reduction($reductionType, $reductionValue);
+        $this->includeTax = $includeTax;
+        $this->price = new Number((string) $price);
+        $this->fromQuantity = $fromQuantity;
+    }
+
+    /**
+     * @return ProductId
+     */
+    public function getProductId(): ProductId
+    {
+        return $this->productId;
+    }
+
+    /**
+     * @return Reduction
+     */
+    public function getReduction(): Reduction
+    {
+        return $this->reduction;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isIncludeTax(): bool
+    {
+        return $this->includeTax;
+    }
+
+    /**
+     * @return Number
+     */
+    public function getPrice(): Number
+    {
+        return $this->price;
+    }
+
+    /**
+     * @return int
+     */
+    public function getFromQuantity(): int
+    {
+        return $this->fromQuantity;
+    }
+
+    /**
+     * @return DateTime|null
+     */
+    public function getDateTimeFrom(): ?DateTime
+    {
+        return $this->dateTimeFrom;
+    }
+
+    /**
+     * @param DateTime|null $dateTimeFrom
+     */
+    public function setDateTimeFrom(?DateTime $dateTimeFrom): void
+    {
+        $this->dateTimeFrom = $this->createDateTime($dateTimeFrom);
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getShopGroupId(): ?int
+    {
+        return $this->shopGroupId;
+    }
+
+    /**
+     * @param int $shopGroupId
+     */
+    public function setShopGroupId(int $shopGroupId): void
+    {
+        $this->shopGroupId = $shopGroupId;
```

Comment:
In some cases i find VO's hard to use(talking about Id), because it cannot have valid value with id "0". Like in edit actions, where property value can be "0", but cannot be null
(because it has different cases: 
null - when value hasn't changed and shouldn't be updated,
"0" - when value was reset intentionally and should be updated.). Which is similar to this one. Although I could set those Ids to VO's after checking if they aren't null because its command for creation in this case.

Overall as far as I understand it depends from how much domain validation rules property contains. If it has some constants and/or couple validation methods then it suppose to be a VO.
Although i'm not really sure why is it still worth using all those "id" objects, after typehints are available. Seems a little overhead, but maybe it makes it harder to do mistakes.

Reply:
I guess it's more protective to use VO indeed, as for the "0" use case I don't think an ID should be allowed to have a 0 value
Or is it used when you want to clean an association for example? To clear a foreign key?
In which case I understand the need to have a difference between 0 and null, but then the VO object should precisely be used to validate or not these behaviour (not all IDs/associations can be reset)

Reply:
> Or is it used when you want to clean an association for example? To clear a foreign key?
yep.
> VO object should precisely be used to validate or not these behaviour
have any suggestion? I've already seen couple cases like this, but just used scalar values there

Reply:
Can be refactored after QA check 👍